### DISCUSSION TITLE
lc/1/

### DISCUSSION THREAD

**Discussion by giscus** [doocs/leetcode#2250] [Announcements]
# lc/1/

多种编程语言实现 LeetCode、《剑指 Offer（第 2 版）》、《程序员面试金典（第 6 版）》题解

https://leetcode.doocs.org/lc/1/

**Comment by yanglbme** [comment]
最简单高效的解法，就是用哈希表

**Comment by 15807076130** [comment]
其他解法
解法一：对于每个元素nums[i]，遍历右侧元素是否存在target-nums[i]。时间复杂度o(n^2)
解法二：先对数组排序，然后双指针分别指向数组开头和结尾，两元素相加，若和>target，则右指针左移，若和<target，则左指针右移，直至找到答案。时间复杂度取决于排序方法

**Reply by yanglbme** [reply]
这两种解法可以的，不过复杂度高于哈希表的解法

**Comment by zhuhui132** [comment]
zhuhui  2024.06.04 打卡

**Reply by yanglbme** [reply]
欢迎评论区打卡留言

**Comment by rain84** [comment]
Additionally to hash-table you can use approach with 2 pointers.
Formally time complexity will be the same O(n), but technically it should be 2 times faster.
But unfortunately it also comes with code duplication.. 🤷🏻

```ts
function twoSum(nums: number[], target: number): number[] {
  const map: Record<number, number> = {}
  const n = nums.length

  for (let l = 0, r = n - 1; l <= r; l++, r--) {
    let v = nums[l]
    let diff = target - v

    if (map[diff] !== undefined) return [map[diff], l]
    map[v] = l

    v = nums[r]
    diff = target - v

    if (map[diff] !== undefined) return [map[diff], r]
    map[v] = r
  }

  return []
}
```

**Comment by ghost** [comment]
我直接return num1 + num2，然后你告诉我还能这么些

**Comment by kilitezh** [comment]
小白请教一下，在java的实现中为什么for的循环条件可以设置为空，这样不会导致无限循环吗

**Reply by yanglbme** [reply]
因为题目保证一定存在一个有效答案。所以一定会 return，不会无限循环

**Comment by 15807076130** [comment]
java有两种for循环。这里只讲和C语言格式一致的for循环。你只需要知道for循环代码块的执行顺序就可以了。确实可以设空。按顺序执行能退出就不是无限循环。
for(1;2;4){
3
}
执行顺序是这样的。


另一种for循环，for(元素变量:集合){
...
}
我不是特别清楚，应该集合为空或者迭代到最后一个元素就可以退出了吧

---Original---
From: ***@***.***&gt;
Date: Mon, May 12, 2025 22:39 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [doocs/leetcode] lc/1/ (Discussion #2250)




 
小白请教一下，在java的实现中为什么for的循环条件可以设置为空，这样不会导致无限循环吗

**Reply by yanglbme** [reply]
题目中，一定会 return，也就退出了循环。

**Comment by phh79** [comment]
```
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> result;
        for (int i = 0; i < nums.size(); i ++) {
            for (int j = i + 1; j < nums.size(); j++) {
                if ((nums[i] + nums[j]) == target) {
                    result = {i, j};
                    return result;
                }
            }
        }
        return result;
    }
};
```
